compiler/pipes/calc-bad-vars.cpp

```diff
@@ -115,13 +115,297 @@ struct FuncCallGraph {
+  function_palette::colors_mask_t colors_mask{0};       // mask of all colors_chain
+
+  void recalc_mask() {
+    colors_mask = 0;
+    for (function_palette::color_t c : colors_chain) {
+      colors_mask |= c;
+    }
+  }
+
+public:
+  CallstackOfColoredFunctions() {
+    stack.reserve(10);
+    index_set.reserve(10);
+  }
+
+  size_t size() const {
+    return stack.size();
+  }
+
+  const std::vector<FunctionPtr> &as_vector() const {
+    return stack;
+  }
+
+  const std::vector<function_palette::color_t> &get_colors_chain() const {
+    return colors_chain;
+  }
+
+  function_palette::colors_mask_t get_colors_mask() const {
+    return colors_mask;
+  }
+
+  void push_back(const FunctionPtr &el) {
+    stack.push_back(el);
+    index_set.insert(el);
+    for (const auto &sep_color : el->colors) {
+      colors_chain.push_back(sep_color);
+    }
+    recalc_mask();
+  }
+
+  void pop_back() {
+    FunctionPtr pop = stack.back();
+    index_set.erase(pop);
+    stack.pop_back();
+    for (const auto &sep_color __attribute__((unused)) : pop->colors) {
+      colors_chain.pop_back();
+    }
+    recalc_mask();
+  }
+
+  bool contains(const FunctionPtr &el) const {
+    return index_set.find(el) != index_set.end();
+  }
+};
+
+// here we check all palette rules like "api has-curl" or "messages-module messages-internals"
+// start from the root, make dfs expanding next_with_colors and check all rules for each dfs chain
+// these chains are left-to-right, e.g. colored f1->f2->f3, we find and check f1, then f1->f2, then f1->f2->f3
+class CheckFunctionsColors {
+  static constexpr int max_shown_errors_count = 10;
+
+  const FuncCallGraph &call_graph;
+  const function_palette::Palette &palette;
+  std::unordered_set<std::string> shown_errors;   // the easiest way to avoid duplicated errors
+
+public:
+  explicit CheckFunctionsColors(const FuncCallGraph &call_graph)
+    : call_graph(call_graph)
+    , palette(G->get_function_palette()) {}
+
+  void check() {
+    CallstackOfColoredFunctions callstack;
+    check_func_dfs(callstack, G->get_main_file()->main_function);
+  }
+
+  void check_func_dfs(CallstackOfColoredFunctions &callstack, const FunctionPtr &func) {
+    callstack.push_back(func);
+    bool was_any_error = false;
+
+    for (const auto &ruleset : palette.get_rulesets()) {
+      for (auto it = ruleset.rbegin(); it != ruleset.rend(); ++it) {  // reverse is important
+        const function_palette::PaletteRule &rule = *it;
+        if (!match_rule(callstack, rule)) {
+          continue;
+        }
+
+        if (rule.is_error()) {
+          show_error_on_rule_broken(callstack, rule);
+          was_any_error = true;
+        }
+        break;
+      }
+    }
+
+    if (!was_any_error && func->next_with_colors != nullptr) {
+      for (FunctionPtr next : *func->next_with_colors) {
+        if (callstack.size() < 50 && !callstack.contains(next)) {
```

Comment:
What would happen if callstack.size() > 50? Just skip it? Shouldn't we fire some warning?

Reply:
Just skip, nothing more